### PR TITLE
Fb1024

### PR DESCRIPTION
**Pull request by nicola-spire**

### CODE REVIEW

File: library/package.json [hunk 1]
```diff
@@ -14,8 +14,10 @@
     "@types/react-router-dom": "4.2.3"
   },
   "dependencies": {
+    "@types/react-transition-group": "^2.0.9",
     "react": "16.2.0",
     "react-dom": "16.2.0",
-    "react-router-dom": "4.2.2"
+    "react-router-dom": "4.2.2",
+    "react-transition-group": "^2.3.1"
```
Comment (kranar) [inline library/package.json]: Use exact versions, ie: "2.3.1"
Do the same for the type definitions.

File: library/package.json [hunk 2]
```diff
@@ -14,8 +14,10 @@
     "@types/react-router-dom": "4.2.3"
   },
   "dependencies": {
+    "@types/react-transition-group": "^2.0.9",
```
Comment (kranar) [inline library/package.json]: This dependency is only needed for development, not production.

File: library/source/components/burger_button.tsx [hunk 3]
```diff
@@ -1,17 +1,87 @@
 import * as React from 'react';
+import {HBoxLayout, Padding, VBoxLayout} from '..';
 
-export interface Properties {}
+/** Specifies the React properties for the burger button. */
+export interface Properties {
 
-export interface State {}
+	/** The parent's onClick */
```
Comment (kranar) [inline library/source/components/burger_button.tsx]: Indentation.

File: library/source/components/burger_button.tsx [hunk 4]
```diff
@@ -1,17 +1,87 @@
...
+  onClick: ()=> void;
+
+  /** The button's color */
+  color: string;
+};
+
+export interface State {
+
```
Comment (kranar) [inline library/source/components/burger_button.tsx]: Remove blank line.

File: library/source/components/burger_button.tsx [hunk 5]
```diff
@@ -1,17 +1,87 @@
...
+  private static TRANSITION_LENGTH = '0.25s';
+
   constructor(properties: Properties) {
     super(properties);
-    this.state = {};
+    this.onToggleMenuOpen = this.onToggleMenuOpen.bind(this);
+  }
+
+  public componentWillReceiveProps(newProps: Properties) {
+  	const colorAnimation = document.getElementById(newProps.color +
```
Comment (kranar) [inline library/source/components/burger_button.tsx]: Indentation.

Comment (kranar) [inline library/source/components/burger_button.tsx]: newProps.color + ???

This won't build.

File: library/source/components/burger_button.tsx [hunk 6]
```diff
@@ -1,17 +1,87 @@
...
+                       dur={BurgerButton.TRANSITION_LENGTH}/>
+              <rect y='0' width='20' height='2' />
+              <rect y='7' width='20' height='2' />
+              <rect y='14' width='20' height='2' />
+            </g>
+          </svg>
+      </button>
+    </HBoxLayout>);
+  }
+  private onToggleMenuOpen(): void {
```
Comment (kranar) [inline library/source/components/burger_button.tsx]: Place an empty line above function definitions.

File: library/source/components/burger_button.tsx [hunk 7]
```diff
@@ -1,17 +1,87 @@
...
+  	this.props.onClick();
```
Comment (kranar) [inline library/source/components/burger_button.tsx]: Indentation.

File: library/source/pages/game_page/game_page.tsx [hunk 8]
```diff
@@ -26,8 +25,9 @@ export class GamePage extends React.Component<Properties, State> {
     super(props);
     this.state = {
       isLoading: true,
-      redirect: null
+      redirect: null,
```
Comment (kranar) [inline library/source/pages/game_page/game_page.tsx]: Remove extraneous comma.

File: library/source/pages/game_page/game_page.tsx [hunk 9]
```diff
@@ -45,6 +45,12 @@ export class GamePage extends React.Component<Properties, State> {
     } else if(this.state.isLoading) {
       return <LoadingPage/>;
     }
-    return <div>Game Page: {this.props.model.name}</div>
+    return (
+      <HBoxLayout height='100%' width='100%'>
+        <SideMenu onClick={this.onClick}/>
+      </HBoxLayout>);
+  }
+  private onClick() {
```
Comment (kranar) [inline library/source/pages/game_page/game_page.tsx]: Empty line above function definitions.

File: library/source/pages/game_page/side_menu.tsx [hunk 10]
```diff
@@ -21,18 +23,166 @@ export interface Properties {
...
-    this.state = {};
+    this.state = {
+      isMenuOpen: true
+    };
+    this.onToggleMenuOpen = this.onToggleMenuOpen.bind(this);
   }
 
   public render(): JSX.Element {
-    return <div>Side Menu!</div>
+
```
Comment (kranar) [inline library/source/pages/game_page/side_menu.tsx]: Remove empty line.

File: library/source/pages/game_page/side_menu.tsx [hunk 11]
```diff
@@ -21,18 +23,166 @@ export interface Properties {
...
+    this.onToggleMenuOpen = this.onToggleMenuOpen.bind(this);
   }
 
   public render(): JSX.Element {
-    return <div>Side Menu!</div>
+    
+    const menuStyle = {
+        backgroundColor: '#FFFFFF',
+        fontFamily: 'Roboto',
+        fontWeight: 'normal',
```
Comment (kranar) [inline library/source/pages/game_page/side_menu.tsx]: Remove extraneous comma.
Indentation is off.

File: library/source/pages/game_page/side_menu.tsx [hunk 12]
```diff
@@ -21,18 +23,166 @@ export interface Properties {
...
+    };
+    const sideMenuStyle = {
+
```
Comment (kranar) [inline library/source/pages/game_page/side_menu.tsx]: Remove blank line, fix indentation.

File: library/source/pages/game_page/side_menu.tsx [hunk 13]
```diff
@@ -21,18 +23,166 @@ export interface Properties {
...
+        transform: 'translateX(0px)',
+        transition: 'transform 0.25s ease-in-out'
+    };
+    const overlayStyle = {
+        position: 'fixed' as 'fixed',
```
Comment (kranar) [inline library/source/pages/game_page/side_menu.tsx]: Indentation is wrong.

File: library/source/pages/game_page/side_menu.tsx [hunk 14]
```diff
@@ -21,18 +23,166 @@ export interface Properties {
...
+        return 'purple'
+      }
+    })();
+    return(
+      <HBoxLayout>
+        <Transition
+            in={this.state.isMenuOpen}
+            timeout={250}
+            component={null}
+          >
```
Comment (kranar) [inline library/source/pages/game_page/side_menu.tsx]: Place the closing > on the same line, ie. component={null}>

File: library/source/pages/game_page/side_menu.tsx [hunk 15]
```diff
@@ -21,18 +23,166 @@ export interface Properties {
...
+          {(status: any) => {
+            return (
+            <HBoxLayout id='overlay' width='100%' height={'100%'}
```
Comment (kranar) [inline library/source/pages/game_page/side_menu.tsx]: Indentation.

File: library/source/pages/game_page/side_menu.tsx [hunk 16]
```diff
@@ -21,18 +23,166 @@ export interface Properties {
...
+      }
+    })();
+    return(
+      <HBoxLayout>
+        <Transition
+            in={this.state.isMenuOpen}
+            timeout={250}
+            component={null}
+          >
+          {(status: any) => {
```
Comment (kranar) [inline library/source/pages/game_page/side_menu.tsx]: Don't mark parameters as any. Just leave it unspecified like (status) => {

File: library/source/pages/game_page/side_menu.tsx [hunk 17]
```diff
@@ -21,18 +23,166 @@ export interface Properties {
...
+          return {opacity: 1};
+        case 'entering':
+          return {opacity: 0.01};
+        case 'exiting':
+          return {opacity: 1};
+        case 'exited':
+          return {opacity: 0.01};
+      }
+    };
+    const slideInTransitionStyle = (status: string): any => {
```
Comment (kranar) [inline library/source/pages/game_page/side_menu.tsx]: Don't mark return types as any, leave it unspecified, ie (status: string) => {

File: library/source/pages/game_page/side_menu.tsx [hunk 18]
```diff
@@ -21,18 +23,166 @@ export interface Properties {
...
+        transition: 'transform 0.25s ease-in-out'
+    };
+    const overlayStyle = {
+        position: 'fixed' as 'fixed',
+        left: '0px',
+        top: '0px',
+        backgroundColor: 'rgba(112, 112, 112, 0.5)',
+        transition: 'opacity 250ms ease-in-out'
+    };
+    const fadeInTransitionStyle = (status: string): any => {
```
Comment (kranar) [inline library/source/pages/game_page/side_menu.tsx]: Don't mark return types as any, leave it unspecified.

File: library/source/pages/game_page/side_menu.tsx [hunk 19]
```diff
@@ -21,18 +23,166 @@ export interface Properties {
...
+          return {transform: 'translateX(-200px)'};
+        case 'exiting':
+          return {transform: 'translateX(0px)'};
+        case 'exited':
+          return {transform: 'translateX(-200px)'};
+      }
+    };
+    const buttonColor = (() => {
+      if(this.state.isMenuOpen) {
+        return 'white'
```
Comment (kranar) [inline library/source/pages/game_page/side_menu.tsx]: Missing semi colon.

File: library/source/pages/game_page/side_menu.tsx [hunk 20]
```diff
@@ -21,18 +23,166 @@ export interface Properties {
...
+      } else {
+        return 'purple'
```
Comment (kranar) [inline library/source/pages/game_page/side_menu.tsx]: Missing semi colon.

File: library/source/pages/game_page/side_menu.tsx [hunk 21]
```diff
@@ -21,18 +23,166 @@ export interface Properties {
...
+            return (
+            <HBoxLayout id='overlay' width='100%' height={'100%'}
+                        style={{...overlayStyle,
+                                ...fadeInTransitionStyle(status)}}/>
+          )}}
+        </Transition>
+        <Transition in={this.state.isMenuOpen}
+                    timeout={250}
+                    component={null}>
+          {(status: any) => {
```
Comment (kranar) [inline library/source/pages/game_page/side_menu.tsx]: Remove the any.

File: library/source/pages/game_page/side_menu.tsx [hunk 22]
```diff
@@ -21,18 +23,166 @@ export interface Properties {
...
+            return (
+          <HBoxLayout height={this.state.isMenuOpen?'100%': null}
```
Comment (kranar) [inline library/source/pages/game_page/side_menu.tsx]: Indentation. Also put spaces around the ? and :

File: library/source/pages/game_page/side_menu.tsx [hunk 23]
```diff
@@ -21,18 +23,166 @@ export interface Properties {
...
+                      style={{...sideMenuStyle,
+                              ...slideInTransitionStyle(status)}}>
+            <VBoxLayout id='side-menu' style={menuStyle} width='200px'
+                        height='100%'>
+                <Padding size='20px' />
```
Comment (kranar) [inline library/source/pages/game_page/side_menu.tsx]: Indentation

Comment (kranar) [inline library/source/pages/game_page/side_menu.tsx]: No space before the />

File: library/source/pages/game_page/side_menu.tsx [hunk 24]
```diff
@@ -21,18 +23,166 @@ export interface Properties {
...
+                <HBoxLayout width='100%' height='100%'>
+                  <Padding size='17px' />
```
Comment (kranar) [inline library/source/pages/game_page/side_menu.tsx]: No space before the />

Make the fix for any further instance of this.

File: library/source/pages/game_page/side_menu.tsx [hunk 25]
```diff
@@ -21,18 +23,166 @@ export interface Properties {
...
+                    <Padding size='20px'/>
+                  <a className='sidebar-link'
+                     onClick={() => this.props.onClick(SideMenuItem.EXIT)}>
+                     Exit
+                  </a>
+                  <Padding size='40px'/>
+                </VBoxLayout>
+              </HBoxLayout>
+            </VBoxLayout>
+             <HBoxLayout>
```
Comment (kranar) [inline library/source/pages/game_page/side_menu.tsx]: Indentation

File: library/source/pages/game_page/side_menu.tsx [hunk 26]
```diff
@@ -21,18 +23,166 @@ export interface Properties {
...
+              <VBoxLayout>
+                <Padding size='3px'/>
+                <BurgerButton color={buttonColor}
+                              onClick={this.onToggleMenuOpen} />
+              </VBoxLayout>
+            </HBoxLayout>
+          </HBoxLayout>);
+          }}
+        </Transition>
+
```
Comment (kranar) [inline library/source/pages/game_page/side_menu.tsx]: Remove blank line.

File: library/source/pages/game_page/side_menu.tsx [hunk 27]
```diff
@@ -21,18 +23,166 @@ export interface Properties {
...
+                              onClick={this.onToggleMenuOpen} />
+              </VBoxLayout>
+            </HBoxLayout>
+          </HBoxLayout>);
+          }}
+        </Transition>
+        
+      </HBoxLayout>);
+  }
+  private onToggleMenuOpen(): void {
```
Comment (kranar) [inline library/source/pages/game_page/side_menu.tsx]: Blank line before definitions.

File: tests/game_page_tester/source/index.tsx [hunk 28]
```diff
@@ -0,0 +1,7 @@
+import * as React from 'react';
+import * as ReactDOM from 'react-dom';
+import * as sudoku from 'sudoku';
+const board = new sudoku.Board();
```
Comment (kranar) [inline tests/game_page_tester/source/index.tsx]: Put an empty line after the import section.

File: tests/game_page_tester/webpack.config.js [hunk 29]
```diff
@@ -0,0 +1,34 @@
...
+        test: /\.tsx?$/,
+        loader: 'ts-loader'
+      },
+      {
+        enforce: 'pre',
+        test: /\.js$/,
+        loader: 'source-map-loader'
+      }
+    ]
+  },
```
Comment (kranar) [inline tests/game_page_tester/webpack.config.js]: Remove the extraneous comma.

File: library/source/pages/game_page/side_menu.tsx [hunk 30]
```diff
@@ -42,23 +42,22 @@ export class SideMenu extends React.Component<Properties, State> {
...
-        backgroundColor: 'rgba(112, 112, 112, 0.5)',
-        transition: 'opacity 250ms ease-in-out'
+      position: 'fixed' as 'fixed',
+      left: '0px',
+      top: '0px',
+      backgroundColor: 'rgba(112, 112, 112, 0.5)',
+      transition: 'opacity 250ms ease-in-out'
     };
-    const fadeInTransitionStyle = (status: string): any => {
+    const fadeInTransitionStyle = (status: string): object => {
```
Comment (kranar) [inline library/source/pages/game_page/side_menu.tsx]: Remove the : object. just have it as so:

(status: string) => {

File: library/source/pages/game_page/side_menu.tsx [hunk 31]
```diff
@@ -70,7 +69,7 @@ export class SideMenu extends React.Component<Properties, State> {
           return {opacity: 0.01};
       }
     };
-    const slideInTransitionStyle = (status: string): any => {
+    const slideInTransitionStyle = (status: string): object => {
```
Comment (kranar) [inline library/source/pages/game_page/side_menu.tsx]: Remove object/return type.

File: library/source/pages/game_page/side_menu.tsx [hunk 32]
```diff
@@ -84,102 +83,104 @@ export class SideMenu extends React.Component<Properties, State> {
...
+                                  onClick={this.onToggleMenuOpen}/>
+                  </VBoxLayout>
+                </HBoxLayout>
+              </HBoxLayout>);
           }}
         </Transition>
-        
       </HBoxLayout>);
   }
   private onToggleMenuOpen(): void {
```
Comment (kranar) [inline library/source/pages/game_page/side_menu.tsx]: Place a blank line above the function declaration.

**Comment by kranar** [comment]
I will close the pull request for now until the animation changes are made. Overall it's really good though.

**Comment by nicola-spire** [comment]
Sounds great!
That was a fun discussion.

**Pull request closed without merging**